File: NC/longest-substring-without-duplicates.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        if len(s) <= 1:
            return len(s)

        ans = 1

        l, r = 0, 1

        count = {s[0]: 1}

        while r < len(s):
            count[s[r]] = count.get(s[r], 0) + 1

            if count[s[r]] > 1:
                while count[s[r]] != 1:
                    count[s[l]] -= 1
                    l += 1

            ans = max(ans, r-l+1)
            r += 1

        return ans

File: NC/longest-substring-without-duplicates.cpp
class Solution
{
public:
    int lengthOfLongestSubstring(string s)
    {
        unordered_map<char, int> charCount;

        int ans = 0;
        if (s.size() < 1)
        {
            return 0;
        }
        else
        {
            ans = 1;
            charCount[s[0]]++;
        }

        for (int i = 0, j = 1; j < s.length(); j++)
        {
            char currentChar = s[j];
            charCount[currentChar]++;
            while (charCount[currentChar] > 1)
            {
                charCount[s[i]]--;
                i++;
            }
            ans = max(ans, j - i + 1);
        }
        return ans;
    }
};
